File: app/src/main/java/com/example/kocja/rabbiter_reworked/activities/viewEntry.java
package com.example.kocja.rabbiter_reworked.activities;

import android.content.Intent;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.example.kocja.rabbiter_reworked.GsonManager;
import com.example.kocja.rabbiter_reworked.HttpManager;
import com.example.kocja.rabbiter_reworked.R;
import com.example.kocja.rabbiter_reworked.databases.Entry;
import com.example.kocja.rabbiter_reworked.databases.Events;
import com.example.kocja.rabbiter_reworked.fragments.HistoryFragment;
import com.example.kocja.rabbiter_reworked.fragments.viewEntryData;
import com.google.gson.Gson;

import java.util.UUID;


/**
 * Created by kocja on 27/01/2018.
 */

public class viewEntry extends AppCompatActivity {
    private Entry mainEntry;
    private viewEntryData mainEntryFragment;
    private UUID mainEntryUUID;
    private boolean dataChanged = false;
    private ImageView mainEntryView;
    private Gson gson;

    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_view_entry);
        Intent currentIntent = getIntent();

        gson = GsonManager.getGson();
        mainEntry = new Entry();
        mainEntry.entryID=(UUID) currentIntent.getSerializableExtra("entryID");
        mainEntryView = findViewById(R.id.mainEntryView);
        Intent viewLargerImage = new Intent(this,largerMainImage.class);
        mainEntryView.setOnClickListener(view -> {
            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP){
                mainEntryView.setTransitionName("closerLook");
                ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(this,mainEntryView,"closerLook");
                startActivity(viewLargerImage,options.toBundle());
            }
            else{
                startActivity(viewLargerImage);
            }

        });
        Log.d("log",gson.toJson(mainEntry));
        HttpManager.postRequest("seekSingleEntry",gson.toJson(mainEntry), (response,bytes) -> {
            Log.d("res",response);
            Entry entry = gson.fromJson(response,Entry[].class)[0];
            viewLargerImage.putExtra("imageURI",entry.entryPhLoc);

            mainEntry = entry;
            mainEntryFragment = (viewEntryData) getSupportFragmentManager().findFragmentById(R.id.mainEntryFragment);

            RecyclerView historyView = findViewById(R.id.upcomingAdapter);

            if(entry.chooseGender.equals(getString(R.string.genderMale))){
                HistoryFragment.maleParentOf(viewEntry.this, entry.entryName,historyView,viewEntry.this);
            }
            else {
                HistoryFragment.setPastEvents(viewEntry.this,entry.entryName,historyView);
            }

            mainEntryFragment.setData(entry);

            HttpManager.postRequest("searchForImage", gson.toJson(entry.entryPhLoc), (response1, bytes1) -> {
                entry.entryBitmap = BitmapFactory.decodeByteArray(bytes1,0, bytes1.length);
                this.runOnUiThread(() -> Glide.with(this).load(entry.entryBitmap).into(mainEntryView));
            });


            if(entry.isMerged){
                ImageView mergedView = findViewById(R.id.mergedView);
                mergedView.setOnClickListener(view -> {
                    Intent startMergedMain = new Intent(viewEntry.this,viewEntry.class);
                    startMergedMain.putExtra("entryID",entry.mergedEntry);
                    ActivityOptionsCompat compat = ActivityOptionsCompat.makeSceneTransitionAnimation(viewEntry.this,mainEntryView,"mergedName");
                    startActivity(startMergedMain,compat.toBundle());
                });
                mergedView.setVisibility(View.VISIBLE);
                HttpManager.postRequest("searchForImage", gson.toJson(entry.mergedEntryPhLoc), (response1, bytes1) -> {
                    entry.mergedEntryBitmap = BitmapFactory.decodeByteArray(bytes1,0, bytes1.length);
                    viewEntry.this.runOnUiThread(() -> Glide.with(viewEntry.this).load(entry.mergedEntryBitmap).into(mergedView));
                });

            }
        });
    }
    public void onActivityResult(int requestCode,int resultCode,Intent data){
        if(requestCode == addEntryActivity.EDIT_EXISTING_ENTRY && resultCode == RESULT_OK){
            HttpManager.postRequest("seekSingleEntry", gson.toJson(mainEntry), (response,bytes) -> {
                Entry entry = gson.fromJson(response,Entry[].class)[0];
                mainEntryFragment.setData(entry);
                Glide.with(viewEntry.this)
                        .load(entry.entryPhLoc)
                        .into(mainEntryView);
            });
            dataChanged = true;
        }
    }
    public void onBackPressed(){
        if(dataChanged){
            setResult(RESULT_OK);
            supportFinishAfterTransition();
            //finish();
        }
        else{
            super.onBackPressed();
        }
    }
    public boolean onCreateOptionsMenu(Menu menu){
        getMenuInflater().inflate(R.menu.menu_view_entry_data,menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.editEntry) {
            Intent startEditProc = new Intent(viewEntry.this, addEntryActivity.class);
            startEditProc.putExtra("getMode", addEntryActivity.EDIT_EXISTING_ENTRY);
            startEditProc.putExtra("entryEdit", mainEntry.entryID);
            startActivityForResult(startEditProc, addEntryActivity.EDIT_EXISTING_ENTRY);

        } else if (id == R.id.deleteEntry) {
            AlertDialog.Builder assureDeletion = new AlertDialog.Builder(viewEntry.this)
                    .setTitle(R.string.confirmDeletion)
                    .setPositiveButton(R.string.confirm, (dialogInterface, i) -> {
                        HttpManager.postRequest("seekEventsName", gson.toJson(mainEntry.entryName), (response,bytes) -> {
                            Events[] events = gson.fromJson(response,Events[].class);
                             for (Events event : events) {
                                HttpManager.postRequest("deleteEvent", gson.toJson(event.eventUUID), (response1,bytes1) -> {});
                            }
                            HttpManager.postRequest("deleteEntry", gson.toJson(mainEntry.entryID), (response1,bytes1) -> {});
                            setResult(RESULT_OK);
                            finish();
                        });
                    })
                    .setNegativeButton(R.string.decline, (dialogInterface, i) -> dialogInterface.cancel());
            assureDeletion.show();

        }
        else if (id == R.id.entryStats) {
            Intent startStatActivity = new Intent(getApplicationContext(), viewEntryStats.class);
            startStatActivity.putExtra("entryUUID", mainEntry.entryID);
            startActivity(startStatActivity);

        }
        else if(id == R.id.showMerged){
            if(mainEntry.isMerged){
                Intent showMerged = new Intent(getApplicationContext(),viewEntry.class);
                showMerged.putExtra("entryID",mainEntry.mergedEntry);
                startActivity(showMerged);
            }
            else{
                AlertDialog.Builder alertNotMerged = new AlertDialog.Builder(this)
                        .setTitle("Oops")
                        .setMessage("your entry is not merged")
                        .setPositiveButton("OK", (dialogInterface, i) -> dialogInterface.cancel());
                alertNotMerged.show();
            }

        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/example/kocja/rabbiter_reworked/activities/viewEntry.kt
package com.example.kocja.rabbiter_reworked.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.core.app.ActivityOptionsCompat
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity

import com.bumptech.glide.Glide
import com.example.kocja.rabbiter_reworked.R
import com.example.kocja.rabbiter_reworked.databases.Entry
import com.example.kocja.rabbiter_reworked.databases.Entry_Table
import com.example.kocja.rabbiter_reworked.databases.Events
import com.example.kocja.rabbiter_reworked.databases.Events_Table
import com.example.kocja.rabbiter_reworked.fragments.HistoryFragment
import com.example.kocja.rabbiter_reworked.fragments.viewEntryData
import com.raizlabs.android.dbflow.sql.language.SQLite
import kotlinx.android.synthetic.main.activity_view_entry.*
import kotlinx.android.synthetic.main.upcoming_history_fragment_layout.*

import java.util.UUID

/**
 * Created by kocja on 27/01/2018.
 */

class viewEntry : AppCompatActivity() {
    private var mainEntry: Entry? = null
    private var mainEntryFragment: viewEntryData? = null
    private var mainEntryUUID: UUID? = null
    private var dataChanged = false
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_view_entry)
        val currentIntent = intent
        mainEntryUUID = currentIntent.getSerializableExtra("entryID") as UUID
        val viewLargerImage = Intent(this, largerMainImage::class.java)
        mainEntryView!!.setOnClickListener {
            mainEntryView!!.transitionName = "closerLook"
            val options = ActivityOptionsCompat.makeSceneTransitionAnimation(this, mainEntryView!!, "closerLook")
            startActivity(viewLargerImage, options.toBundle())

        }
        SQLite.select()
                .from(Entry::class.java)
                .where(Entry_Table.entryID.eq(mainEntryUUID))
                .async()
                .querySingleResultCallback { _, entry ->
                    viewLargerImage.putExtra("imageURI", entry?.entryPhLoc)

                    mainEntry = entry
                    mainEntryFragment = supportFragmentManager.findFragmentById(R.id.mainEntryFragment) as viewEntryData?


                    if (entry!!.chooseGender == getString(R.string.genderMale)) {
                        HistoryFragment.maleParentOf(this, entry.entryName, upcomingAdapter, this@viewEntry)
                    } else {
                        HistoryFragment.setPastEvents(this, entry.entryName, upcomingAdapter)
                    }

                    mainEntryFragment!!.setData(entry)

                    Glide.with(this@viewEntry).load(entry.entryPhLoc).into(mainEntryView!!)

                    if (entry.isMerged) {
                        mergedView.setOnClickListener {
                            val startMergedMain = Intent(this, viewEntry::class.java)
                            startMergedMain.putExtra("entryID", entry.mergedEntry!!.entryID)
                            val compat = ActivityOptionsCompat.makeSceneTransitionAnimation(this, mainEntryView!!, "mergedName")
                            startActivity(startMergedMain, compat.toBundle())
                        }
                        mergedView.visibility = View.VISIBLE
                        Glide.with(this).load(entry.mergedEntryPhLoc).into(mergedView)
                    }
                }.execute()
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == AddEntryActivity.EDIT_EXISTING_ENTRY && resultCode == Activity.RESULT_OK) {
            SQLite.select()
                    .from(Entry::class.java)
                    .where(Entry_Table.entryID.eq(mainEntryUUID))
                    .async()
                    .querySingleResultCallback { _, entry ->
                        mainEntryFragment!!.setData(entry)
                        Glide.with(this)
                                .load(entry?.entryPhLoc)
                                .into(mainEntryView!!)
                    }.execute()
            dataChanged = true

        }
    }

    override fun onBackPressed() {
        if (dataChanged) {
            setResult(Activity.RESULT_OK)
            supportFinishAfterTransition()
            //finish();
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_view_entry_data, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.editEntry) {
            val startEditProc = Intent(this@viewEntry, AddEntryActivity::class.java)
            startEditProc.putExtra("getMode", AddEntryActivity.EDIT_EXISTING_ENTRY)
            startEditProc.putExtra("entryEdit", mainEntry!!.entryID)
            startActivityForResult(startEditProc, AddEntryActivity.EDIT_EXISTING_ENTRY)

        } else if (id == R.id.deleteEntry) {
            val assureDeletion = AlertDialog.Builder(this@viewEntry)
                    .setTitle(R.string.confirmDeletion)
                    .setPositiveButton(R.string.confirm) { _, _ ->
                        SQLite.select()
                                .from(Events::class.java)
                                .where(Events_Table.name.eq(mainEntry!!.entryName))
                                .async()
                                .queryListResultCallback { _, tResult ->
                                    for (event in tResult) {
                                        event.delete()
                                    }
                                    mainEntry!!.delete()
                                    setResult(Activity.RESULT_OK)
                                    finish()
                                }.execute()
                    }
                    .setNegativeButton(R.string.decline) { dialogInterface, _ -> dialogInterface.cancel() }
            assureDeletion.show()

        } else if (id == R.id.entryStats) {
            val startStatActivity = Intent(applicationContext, viewEntryStats::class.java)
            startStatActivity.putExtra("entryUUID", mainEntry!!.entryID)
            startActivity(startStatActivity)

        } else if (id == R.id.showMerged) {
            if (mainEntry!!.isMerged) {
                val showMerged = Intent(applicationContext, viewEntry::class.java)
                showMerged.putExtra("entryID", mainEntry!!.mergedEntry!!.entryID)
                startActivity(showMerged)
            } else {
                val alertNotMerged = AlertDialog.Builder(this)
                        .setTitle("Oops")
                        .setMessage("your entry is not merged")
                        .setPositiveButton("OK") { dialogInterface, _ -> dialogInterface.cancel() }
                alertNotMerged.show()
            }

        }
        return super.onOptionsItemSelected(item)
    }
}
